README.md
# Overview

This package implements the simulation in 20.2 of _Genetics From Genes to Genomes, 5th_.



# Installation

```r
if (!('remotes' %in% .packages(T))) install.packages('remotes');
remotes::install_github('dongzhuoer/pgmcs');
```



# For developers

1. TO DO: replace **plyr** with **purr**

```r
remotes::install_github('dongzhuoer/pgmcs')

# https://www.shinyapps.io/admin/#/tokens
# comment `RETICULATE_PYTHON` in `~/.Renvrion`, I don't know why but it just cause error
rsconnect::deployApp('inst/shiny/pgmcs/', appName= 'population-genetics-monte-carlo-simulation', forceUpdate = T)
rsconnect::deployApp('inst/shiny/pgmcs2/',  appName= 'population-genetics-monte-carlo-simulation2', forceUpdate = T)
```


NEWS.md
# pgmcs 0.3.2

## Minor improvements and fixes

1. shorten name of shiny App folder, avoid too long path for `remotes::install_git('.')`


# pgmcs 0.3.1

## Minor improvements and fixes

1. add `RNGversion('3.5.2')` to make vignette and shiny example reproduciable  
   > The default method for generating from a discrete uniform distribution (used in `sample()`, for instance) has been changed.  
   > --- [CHANGES IN R 3.6.0](https://cran.r-project.org/doc/manuals/r-devel/NEWS.html)


# pgmcs 0.3.0

## New features

1. improve documentation with pkgdown
1. add testthat


# pgmcs 0.2.1

## Minor improvements and fixes

1. fix package dependencies  (#1)


# pgmcs 0.2.0

## Breaking changes

1. Rename to `pgmcs`. (Thanks to someone for releasing a package with the same name to CRAN)

## New features

1. a new Shiny app which is much faster (Calculation is only performed after user has pressed the "Run" button).   

## Minor improvements and fixes

1. Polish app's style, make it into a 1366\*768 screen. (It should can if the user don't have too much thing up and down in the screen, such as toolbar, menubar. Anyway I don't make sure since I don't have a 1366\*768 laptop.)


# pgs 0.1.0

## New features

1. `population_simulation()` for one run
1. `demo_population_simulation()` for multiple runs at the same time
1. a Shiny app including demos. However, the speed is kinds of _slow_ since the result is updated in real time (e.g., when you slide a value from 10 to 1000, 20, 50, 120, 270, etc are also calculated) 


tests/testthat/setup.R
testthat::context('Initialize environment')
# refer to http://testthat.r-lib.org/reference/test_dir.html#test-files



suppressWarnings(RNGversion('3.5.2'))


tests/testthat/test-population-simulation.R
testthat::context('Testing population_simulation.R')
setwd(here::here(''))  # workspace is reset per file



testthat::test_that('population_simulation()', {
    set.seed(0)
    
    testthat::expect_identical(
        population_simulation(10, c(0, 10, 0)),
        c(0.5, 0.55, 0.7, 0.7, 0.65, 0.8, 1, 1, 1, 1)
    ) 
});


tests/testthat/test-demo-population-simulation.R
testthat::context('Testing demo_population_simulation.R')
setwd(here::here(''))  # workspace is reset per file



testthat::test_that('demo_population_simulation()', {
    testthat::expect_true(T);

    set.seed(13);
    demo_population_simulation(6, 30, c(0, 10, 0), c(1, 1, 1), "Small population: Lots of genetic drift");

    set.seed(0);
    demo_population_simulation(6, 30, c(0, 500, 0), c(1, 1, 1), "Large population: Little genetic drift");

    set.seed(57);
    demo_population_simulation(6, 30, c(0, 500, 0), c(1,1,0.8), "Natural selection")
});


R/population-simulation.R
#' @title simulate genetic drift in a finite population
#'
#' @description simulate genetic drift in a finite population for one gene with two alleles
#'
#' @param ngeneration integer scalar. Number of generations
#' @param population integer vector. How many dominant homozygotes, heterozygotes, and recessive homozygotes do the initial population contain. For example, `c(2, 5, 3)` means 2 AA, 5 Aa and 3 aa.
#' @param fitness numeric vector. The fitness value of AA, Aa and aa. `c(1, 1, 1)` means no natural selection.
#'
#' @return numeric vector. The frequency of allele A.
#'
#' @examples
#' population_simulation(30, c(0, 10, 0))
#'
#' @export
population_simulation <- function(ngeneration, population, fitness = c(1, 1, 1)) {
    population <- rep(c('AA', 'Aa', 'aa'), population);
    n <- length(population);             # population size
    A.frequency <- numeric(ngeneration); # to store allele frequency

    for (i in seq_len(ngeneration)) {
        A.frequency[i] = mean(ifelse(population == 'AA', 1, ifelse(population == 'aa', 0, 0.5)))

        ## function for generate m offspring
        ## m must >= 1
        reproduce <- function(population, m) {
            ## we don't consider gender here, so mother and father can be the same individual
            gamete <- function(parent) {
                ifelse(parent == 'AA', 'A', ifelse(parent == 'aa', 'a', sample(c('A', 'a'), m, T)))
            }

            parent = sample(population, 2*m, T);

            father = parent[seq(1, 2*m, 2)];
            mather = parent[seq(2, 2*m, 2)];

            sperm = gamete(father);
            ovum = gamete(mather);

            paste0(sperm, ovum);
        }

        # generate `n` new births to replace the formal generation
        population = reproduce(population, n);

        # selection
        if (!identical(fitness, c(1, 1, 1))) {
            # if Wbb is 0.98, we randomly eliminates 2% of bb individuals created in each generation, and replaces them with off spring from a new mating of the parental generation.

            nAA = sum(population == 'AA');
            naa = sum(population == 'aa');
            nAa = n - nAA - naa; # the population size is constant, and heterozygous may be 'Aa' or 'aA'

            new.nAA = round(nAA * fitness[1]);
            new.nAa = round(nAa * fitness[2]);
            new.naa = round(naa * fitness[3]);

            ndead = n - new.nAA - new.nAa - new.naa;

            if (ndead > 0) {
                population <- c(rep('AA', new.nAA), rep('Aa', new.nAa), rep('aa', new.naa),
                                reproduce(population, ndead));
            }
        }
    }

    return(A.frequency);
}


R/demo-population-simulation.R
#' @title run several population simulations
#'
#' @description 
#' run several population simulations and visualize the result, refer to `vignette('pgmcs')` for examples
#'
#' @details the underlying workhorse is [population_simulation()]
#'
#' @param nrun integer scalar. Number of simulation runs.
#' @inheritParams population_simulation
#' @param title string. Plot title.
#'
#' @return an ggplot object
#'
#' @section Shiny app:
#' `shiny::runApp(system.file('shiny', package = 'pgs'))`
#' 
#' @export
demo_population_simulation <- function(nrun, ngeneration, population, fitness, title) {
    plyr::ldply(1:nrun, function(i){
        data.frame(irun = i, igeneration = seq_len(ngeneration),
                   A.frequency = population_simulation(ngeneration, population, fitness))
    }) -> df;

    g <- ggplot2::ggplot(df, ggplot2::aes(x = igeneration, y = A.frequency, color = factor(irun))) +
        ggplot2::geom_line() +
        ggplot2::ylim(0,1) +
        ggplot2::scale_color_discrete(guide = ggplot2::guide_legend(title = 'run')) +
        ggplot2::labs(title = title, x = 'Generations from start (generation 0)', y = 'Frequency of allele A') +
        ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, size = 15))

    if (ngeneration <= 100) g = g + ggplot2::geom_point();
    if (nrun > 10) g = g + ggplot2::theme(legend.position = 'none')
    g;
}


inst/shiny/pgmcs/app.R
library(shiny)
library(pgmcs)

suppressWarnings(RNGversion('3.5.2'))


ui <- fluidPage(
    titlePanel('Population Genetics Monte Carlo Simulation, dynamic mode'),

    includeCSS('style.css'),

    fluidRow(
        column(3, h3('')),
        column(3, h3('AA'), style = 'font-size:150%;text-align:center'),
        column(3, h3('Aa'), style = 'font-size:150%;text-align:center'),
        column(3, h3('aa'), style = 'font-size:150%;text-align:center')
    ),

    fluidRow(
        column(3, p('initial poppulation', style = 'font-size:150%;color:steelblue;text-align:center')),
        column(3, sliderInput('nAA', NULL, value = 0, min = 0, max = 500, step = 1)),
        column(3, sliderInput('nAa', NULL, value = 0, min = 0, max = 500, step = 1)),
        column(3, sliderInput('naa', NULL, value = 0, min = 0, max = 500, step = 1))
    ),

    fluidRow(
        column(3, p('fitness value', style = 'font-size:150%;color:steelblue;text-align:center')),
        column(3, sliderInput('wAA', NULL, value = 0, min = 0, max = 1, step = 0.01)),
        column(3, sliderInput('wAa', NULL, value = 0, min = 0, max = 1, step = 0.01)),
        column(3, sliderInput('waa', NULL, value = 0, min = 0, max = 1, step = 0.01))
    ),

    fluidRow(
        column(2, sliderInput('nrun', h4('run'), value = 0, min = 1, max = 100, step = 1)),
        column(3, sliderInput('ngeneration', h4('generation'), value = 0, min = 5, max = 1000, step = 5)),
        column(5, textInput('title', h4('title'))),
        column(2, sliderInput('seed', h4('seed'), value = 0, min = 1, max = 100, step = 1))
    ),

    fluidRow(
        column(1, actionButton('run_again', 'Run again', icon('play-circle'))),
        column(1, actionButton('reset', 'Reset', icon('refresh'))),
        column(1, actionButton('fig20.8a', '20.8 (a)')),
        column(1, actionButton('fig20.8b', '20.8 (b)')),
        column(1, actionButton('fig20.13-1', '20.13')),
        column(1, actionButton('fig20.13-2', '20.13')),
        column(6, helpText(HTML('Go to <a href="/population-genetics-monte-carlo-simulation2">static mode</a>. The output will not update in real time, so the speed is much <strong>faster</strong>.')))
    ),

    hr(),

    plotOutput(outputId = 'plot'),

    helpText(strong('Note:'), 'For Figure 20.13, I find it almost impossible to produce a figure which looks like that in the textbook (I run 1000 simulations but hadn\'t found even one). So I adjust 0.98 to 0.97.'),

    helpText(strong('Advanced:'), 'I made several restrictions to relieve the workload of the server, please refer to', a(href = 'https://github.com/dongzhuoer/pgmcs', 'https://github.com/dongzhuoer/pgmcs'), 'if you want to explore more possibilities'),

    helpText(strong('Bug Report:'), 'If you find any bugs, please tell me',
             a(href = 'mailto:dongzhuoer@mail.nankai.edu.cn?subject=Report a bug in pgmcs Shiny app', 'email'),
             a(href = 'https://github.com/dongzhuoer/pgmcs/issues/new', 'GitHub'))
)



server <- function(input, output, session) {
    update_element <- function(nrun, ngeneration, title, seed , nAA, nAa, naa, wAA, wAa, waa){
        updateNumericInput(session, 'nrun', value = nrun);
        updateNumericInput(session, 'ngeneration', value = ngeneration);
        updateNumericInput(session, 'seed', value = seed);
        updateTextInput(session, 'title', value = title);
        updateNumericInput(session, 'nAA', value = nAA);
        updateNumericInput(session, 'nAa', value = nAa);
        updateNumericInput(session, 'naa', value = naa);
        updateNumericInput(session, 'wAA', value = wAA);
        updateNumericInput(session, 'wAa', value = wAa);
        updateNumericInput(session, 'waa', value = waa);
    }

    update_plot <- function() {
        renderPlot({
            set.seed(input$seed);

            nrun <- input$nrun;
            ngeneration <- input$ngeneration;
            pupulation <- c(input$nAA, input$nAa, input$naa);
            fitness <- as.numeric(c(input$wAA, input$wAa, input$waa));
            title <- input$title;

            demo_population_simulation(nrun, ngeneration, pupulation, fitness, title);
        }) -> output$plot;
    }

    reset <- function() {
        update_element(6 ,30, 'Population Genetics Monte Carlo Simulation', 0, 0, 10, 0, 1, 1, 1);
        update_plot();
    }

    example <- function(...) {
        update_element(...);
        update_plot();
    }

    reset();

    observeEvent(input$run_again, {update_plot(); updateNumericInput(session, 'seed', value = sample(1:1000, 1));});
    observeEvent(input$reset, reset());

    observeEvent(input$fig20.8a,
                 example(6, 30, 'Figure 20.8 (a) Small population: Lots of genetic drift',
                         13, 0, 10, 0, 1, 1, 1)
    );

    observeEvent(input$fig20.8b,
                 example(6, 30, 'Figure 20.8 (b) Large population: Little genetic drift',
                         0, 0, 500, 0, 1, 1, 1)
    );

    observeEvent(input$`fig20.13-1`,
                 example(6, 900, 'Figure 20.13 Natural selection together with genetic drift for Small Population',
                         57, 0, 1, 499, 1, 1, 0.97)
    );

    observeEvent(input$`fig20.13-2`,
                 example(6, 900, 'Figure 20.13 Natural selection together with genetic drift for Small Population',
                         885, 0, 1, 499, 1, 1, 0.97)
    );
}



shinyApp(ui = ui, server = server);
